sway-lsp/src/server.rs

```diff
@@ -217,3 +223,310 @@ impl LanguageServer for Backend {
         ))
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use serde_json::json;
+    use std::{env, fs::File, io::Write};
+    use tower::{Service, ServiceExt};
+
+    use super::*;
+    use futures::stream::StreamExt;
+    use tower_lsp::jsonrpc::{self, Request, Response};
+    use tower_lsp::LspService;
+
+    // Simple sway script used for testing LSP capabilites
+    const SWAY_PROGRAM: &str = r#"script;
+
+use std::*;
+
+/// A simple Particle struct
+struct Particle {
+    position: [u64; 3],
+    velocity: [u64; 3],
+    acceleration: [u64; 3],
+    mass: u64,
+}
+
+impl Particle {
+    /// Creates a new Particle with the given position, velocity, acceleration, and mass
+    fn new(position: [u64; 3], velocity: [u64; 3], acceleration: [u64; 3], mass: u64) -> Particle {
+        Particle {
+            position: position,
+            velocity: velocity,
+            acceleration: acceleration,
+            mass: mass,
+        }
+    }
+}
+
+fn main() {
+    let position = [0, 0, 0];
+    let velocity = [0, 1, 0];
+    let acceleration = [1, 1, 0];
+    let mass = 10;
+    let p = ~Particle::new(position, velocity, acceleration, mass);
+}
+"#;
+
+    fn load_test_sway_file(sway_file: &str) -> Url {
+        let file_name = "tmp_sway_test_file.sw";
+        let dir = env::temp_dir().join(file_name);
+        let mut file = File::create(&dir).unwrap();
+        file.write_all(sway_file.as_bytes()).unwrap();
+
+        let path = format!("file:///{}", dir.as_os_str().to_str().unwrap());
```

Comment:
Rather than using `format` here, it might be safer to use [`Url::from_file_path`](https://docs.rs/url/latest/url/struct.Url.html#method.from_file_path).

Reply:
Nice, good spot, I'll change this now to use `Url::from_file_path`